[APINotes] Remove VersionedInfoRole

jrose-apple opened the pull request:
When API notes were first added, they would defer to the header when something was explicitly specified---for example, if the API notes had one set of nullability annotations from before the header was audited. This behavior didn't make sense for versioned API notes, however, where annotations needed for Swift 3 compatibility were intended to override the unversioned information in both the API notes and the header itself.

Now that the versioning of API notes is well-established, simplify things by having the API notes always "win"; that is, an annotation in API notes will always override what's in the headers.

While here, remove the obsolete `FactoryAsInit` key from the API notes dictionary. This never had a good implementation with versioned API notes anyway.

Getting this into a release branch or 'stable' depends on apple/swift#6901 and apple/swift#6907.

jrose-apple committed:
[APINotes] Remove FactoryAsInit (superseded by SwiftName)

This still leaves behind parsing (to emit an error message) as well as
the attribute that used to get generated. Once Swift is no longer
checking for the attribute we can take that out as well.

jrose-apple committed:
Remove the attribute used by the FactoryAsInit API note.

DougGregor commented on lib/APINotes/APINotesWriter.cpp:
```diff
@@ -766,7 +766,7 @@ namespace {
 
     void emitUnversionedInfo(raw_ostream &out, const ObjCMethodInfo &info) {
-      uint8_t payload = info.FactoryAsInit;
+      uint8_t payload = 0;
```
I see that you kept the API notes binary format the same. Thanks!

DougGregor commented on lib/Sema/SemaDecl.cpp:
```diff
@@ -13537,7 +13537,6 @@ Decl *Sema::ActOnTag(Scope *S, unsigned TagSpec, TagUseKind TUK,
 
   if (Attr)
     ProcessDeclAttributeList(S, New, Attr);
-  ProcessAPINotes(New);
```
Why did this go away?

jrose-apple replied:
Whoops! That was supposed to be a separate patch (with a good explanation and a test case). Thanks for catching this!

DougGregor reviewed (changes requested):
Why is it correct to remove the ProcessAPINotes call from ActOnTag? If it is correct, we should separate it out.

jrose-apple committed:
[APINotes] Remove VersionedInfoRole.

When API notes were first added, they would defer to the header when
something was explicitly specified---for example, if the API notes had
one set of nullability annotations from before the header was
audited. This behavior didn't make sense for versioned API notes,
however, where annotations needed for Swift 3 compatibility were
intended to override the unversioned information in both the API notes
and the header itself.

Now that the versioning of API notes is well-established, simplify
things by having the API notes always "win"; that is, an annotation
in API notes will always override what's in the headers.

jrose-apple commented:
Fixed!

DougGregor reviewed (approved):
Thanks!
